Changed DropDown support, Introduced support for several missing elee…

dcorbin opened the pull request:
…nt types;

TableLayout, Row, Cell,
ToggleGroup, ToggleButton,
InputField,
TextElement
Image,
Option

dcorbin committed:
Changed DropDown support, Introduced support for several missing eleent types;

TableLayout, Row, Cell,
ToggleGroup, ToggleButton,
InputField,
TextElement
Image,
Option

dcorbin commented:
I modified how the DropDown was structured. It seemed very inconsistent with the other element types for reasons I could not discern.

dcorbin committed:
Added Defaults and Dropdown as element tags

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -149,6 +156,20 @@ UI = {}
 ---@field scrollSensitivity nil | number @Default 1.0. A factor/multiplier to augment the scroll speed.
 ---@field itemHeight nil | number
 
+---@shape tts__UIDropdownElement : tts__UIInputElementBase
```
Seems like `tts__UIInputElementBase` was removed, yet is still being used. I'm suggesting your reinstate it, however it's worth pointing out nonetheless.

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -96,12 +98,11 @@ UI = {}
 ---@field attributes nil | tts__UIElementBase_Attributes
 ---@field children nil | tts__UIElement[]
 
----@shape tts__UIInputElementBase : tts__UIElementBase
```
Whilst there is indeed an error in the existing types, `tts__UIInputElementBase` should still exist. It's just that it should override the `attributes` field with a new `tts__UIInputElementBase_Attributes` (which itself extends `tts__UIElementBase_Attributes`).

The logic here is that input elements are their own class of elements as they all exhibit common behaviour, namely have an `onValueChanged` callback. Sometimes you may wish to write code that accepts any input element and attaches a `onValueChanged` callback etc.

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -124,7 +125,13 @@ UI = {}
+---@field childControlWidth nil | tts__UIElement_Boolean
+---@field childControlHeight nil | tts__UIElement_Boolean
+
+---@shape tts__UIDropdownElement_Attributes: tts__UIElementBase_Attributes
```
Should inherit from the newly proposed `tts__UIInputElementBase_Attributes`.

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -124,7 +125,13 @@ UI = {}
+---@field childControlHeight nil | tts__UIElement_Boolean
+
+---@shape tts__UIDropdownElement_Attributes: tts__UIElementBase_Attributes
 ---@field onValueChanged tts__UIElement_CallbackFunctionName @Callback name may be optionally followed by "(selectedText)", "(selectedValue)" or "(selectedIndex)". The latter causes the selected Option's index (0-based) to be passed to the callback (but represented as a string). The former two are identical and match the default behavior (no parameter specified) passing the text value of the Opti ...
```
Technically this could be omitted (post-inheritance), however I think it's valuable to keep around for the description.

dcorbin committed:
Fixed panel definition

dcorbin committed:
Fixed setCustomAssets, added GetCustomAssets

dcorbin committed:
Ignore  IDEA folder

dcorbin committed:
Reintroduce tts__UIInputElementBase with tts__UIInputElementBase_Attributes

dcorbin commented:
I _think_ I've addressed the tts__UIInputElementBase/tts__UIInputElementBase_Attributes  issues correctly.
I've also added 3 other small commits, and I'm happy to rebase/squash any the commits if you want, when  the final result has passed muster.

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -170,16 +197,38 @@ UI = {}
+
+---@shape tts__UIToggleGroupElement : tts__UIElementBase
+
+---@shape tts__UIToggleButtonElement_Attributes : tts__UIInputElementBase_Attributes
```
My apologies if this was like this earlier and I missed it. However, Luanalysis requires a blank line between type definition blocks. Otherwise, it gets a bit confused.

In this particular case, seems as `tts__UIToggleButtonElement` doesn't appear to have any attributes of its own. You could just delete this line as well as line 206 (where the `attributes` field is overridden).

dcorbin replied:
No apologies required.  It wasn't that way, I don't think.  It think it was modified post-fixing InputElement things.  There are attributes that just haven't been added yet.    But  I will strike it.

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -170,16 +197,38 @@ UI = {}
+---@field attributes nil | tts__UIPanelElement_Attributes
 ---@field tag "Panel"
 
+---@shape tts__UIOptionElement_Attributes: tts__UIElementBase_Attributes
```
`---@shape Child : Parent` is preferred i.e. spaces surrounding `:`.

dcorbin replied:
Fixed several.

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -1,17 +1,19 @@
 ---@class tts__UI
 UI = {}
 
+---@alias tts__UIEventHandler fun(player: tts__Player, value: string, id: string)
```
Looks like this is a duplicate of `tts__UIElement_CallbackFunction`.

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -96,12 +98,17 @@ UI = {}
 ---@alias tts__UIButtonElement_Transition "None" | "ColorTint" | "SpriteSwap" | "Animation"
 
 ---@shape tts__UIButtonElement_Attributes : tts__UIElementBase_Attributes
+---@field fontSize nil | tts__UIElement_Number  @Default 14
```
There's a stray double space between the type and comment `@`.

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -149,6 +162,20 @@ UI = {}
+
+---@shape tts__UIImageElement_Attributes : tts__UIElementBase_Attributes
+---@field image string
+---@field preserveAspect tts__UIElement_Boolean
```
I suspect this ought to be optional i.e. `nil | tts__UIElement_Boolean`.

dcorbin replied:
Changed.  I just love that the API says the default varies. :)

Benjamin-Dobell commented on tts/UI.def.lua:
```diff
@@ -96,12 +98,17 @@ UI = {}
 ---@field children nil | tts__UIElement[]
 
----@shape tts__UIInputElementBase : tts__UIElementBase
+---@shape tts__UIInputElementBase_Attributes: tts__UIElementBase_Attributes
```
`---@shape Child : Parent` is preferred i.e. spaces surrounding `:`.

dcorbin replied:
Done

Benjamin-Dobell reviewed (changes requested):
My apologies for the back and forth. Looks like I got hung-up on the structural changes and missed a few other things in my initial review.

dcorbin committed:
Address round2 of feedback.  Mostly style
